metron-analytics/metron-profiler-common/src/main/java/org/apache/metron/profiler/hbase/DecodableRowKeyBuilder.java

```diff
@@ -0,0 +1,402 @@
+        groups.add(group);
+      }
+
+      // decode the period
+      long periodId = buffer.getLong();
+      long duration = buffer.getLong();
+      ProfilePeriod period = ProfilePeriod.buildFromPeriod(periodId, duration, TimeUnit.MILLISECONDS);
+
+      return new ProfileMeasurement()
+              .withProfileName(profile)
+              .withEntity(entity)
+              .withGroups(groups)
+              .withProfilePeriod(period);
+
+    } catch(BufferUnderflowException e) {
+      throw new IllegalArgumentException("Unable to decode the row key", e);
+    }
+
+  }
+
+  /**
+   * Builds the 'group' component of the row key.
+   * @param groups The groups to include in the row key.
+   */
+  private byte[] encodeGroups(List<Object> groups) {
+
+    // encode each of the groups and determine their size
+    int lengthOfGroups = 0;
+    List<byte[]> groupBytes = new ArrayList<>();
+    for(Object group : groups) {
+      byte[] groupB = Bytes.toBytes(String.valueOf(group));
+      groupBytes.add(groupB);
+      lengthOfGroups += groupB.length;
+    }
+
+    // encode each of the groups
+    ByteBuffer buffer = ByteBuffer
+            .allocate((Integer.BYTES * (1 + groups.size())) + lengthOfGroups)
+            .order(byteOrder)
+            .putInt(groups.size());
+
+    for(byte[] groupB : groupBytes) {
+      buffer.putInt(groupB.length).put(groupB);
+    }
+
+    return buffer.array();
+  }
+
+  /**
+   * Builds the 'time' portion of the row key
+   * @param period The ProfilePeriod in which the ProfileMeasurement was taken.
+   */
+  private static byte[] encodePeriod(ProfilePeriod period) {
+    return encodePeriod(period.getPeriod());
+  }
+
+  /**
+   * Builds the 'time' portion of the row key
+   * @param period the period
+   */
+  private static byte[] encodePeriod(long period) {
+    return Bytes.toBytes(period);
+  }
+
+  /**
+   * Calculates a salt value that is used as part of the row key.
+   *
+   * The salt is calculated as 'md5(period) % N' where N is a configurable value that ideally
+   * is close to the number of nodes in the Hbase cluster.
+   *
+   * @param period The period in which a profile measurement is taken.
+   */
+  public static byte[] encodeSalt(ProfilePeriod period, int saltDivisor) {
+    return encodeSalt(period.getPeriod(), saltDivisor);
+  }
+
+  /**
+   * Calculates a salt value that is used as part of the row key.
+   *
+   * The salt is calculated as 'md5(period) % N' where N is a configurable value that ideally
+   * is close to the number of nodes in the Hbase cluster.
+   *
+   * @param period The period
+   * @param saltDivisor The salt divisor
+   */
+  public static byte[] encodeSalt(long period, int saltDivisor) {
+    try {
+      // an MD5 is 16 bytes aka 128 bits
+      MessageDigest digest = MessageDigest.getInstance("MD5");
+      byte[] hash = digest.digest(encodePeriod(period));
+      int salt = Bytes.toShort(hash) % saltDivisor;
+      return Bytes.toBytes(salt);
```

Comment:
The other reason to prefix with a hash is in the situation where you have unbalanced keys.  So, the situation where one profile has a lot of data and other profiles don't end up with as much data.  In the case where we want to anticipate splits, having a uniform distribution of keys help because you can just presplit uniformly (as opposed to pre-splitting with a bias for a given key).  Having the uniformly distributed hash prefixing gives a uniformly distributed key distribution, so you can just presplit uniformly.